lib-ee/emqx_ee_connector/src/emqx_ee_connector_sqlserver.erl

```diff
@@ -0,0 +1,526 @@
+    odbc:connect(ConnectStr, Opts).
+
+-spec do_get_status(connection_reference(), time_out()) -> Result :: boolean().
+do_get_status(Conn, RequestTimeout) ->
+    case execute(Conn, <<"SELECT 1">>, RequestTimeout) of
+        {selected, [[]], [{1}]} -> true;
+        _ -> false
+    end.
+
+%%====================================================================
+%% Internal Helper fns
+%%====================================================================
+
+%% TODO && FIXME:
+%% About the connection string attribute `Encrypt`:
+%% The default value is `yes` in odbc version 18.0+ and `no` in previous versions.
+%% And encrypted connections always verify the server's certificate.
+%% So `Encrypt=YES;TrustServerCertificate=YES` must be set in the connection string
+%% when connecting to a server that has a self-signed certificate.
+%% See also:
+%% 'https://learn.microsoft.com/en-us/sql/connect/odbc/
+%%      dsn-connection-string-attribute?source=recommendations&view=sql-server-ver16#encrypt'
+conn_str([], Acc) ->
+    %% we should use this for msodbcsql 18+
+    %% lists:join(";", ["Encrypt=YES", "TrustServerCertificate=YES" | Acc]);
+    lists:join(";", Acc);
+conn_str([{driver, Driver} | Opts], Acc) ->
+    conn_str(Opts, ["Driver=" ++ str(Driver) | Acc]);
+conn_str([{server, Server} | Opts], Acc) ->
+    {Host, Port} = emqx_schema:parse_server(Server, ?SQLSERVER_HOST_OPTIONS),
+    conn_str(Opts, ["Server=" ++ str(Host) ++ "," ++ str(Port) | Acc]);
+conn_str([{database, Database} | Opts], Acc) ->
+    conn_str(Opts, ["Database=" ++ str(Database) | Acc]);
+conn_str([{username, Username} | Opts], Acc) ->
+    conn_str(Opts, ["UID=" ++ str(Username) | Acc]);
+conn_str([{password, Password} | Opts], Acc) ->
+    conn_str(Opts, ["PWD=" ++ str(Password) | Acc]);
+conn_str([{_, _} | Opts], Acc) ->
+    conn_str(Opts, Acc).
+
+%% Sync & Async query with singe & batch sql statement
+-spec do_query(
+    bridge_id(),
+    Query :: {?ACTION_SEND_MESSAGE, map()} | [{?ACTION_SEND_MESSAGE, map()}],
+    ApplyMode ::
+        handover
+        | {handover_async, {?MODULE, do_async_reply, [{ReplyFun :: function(), Args :: list()}]}},
+    state()
+) ->
+    {ok, list()}
+    | {error, {recoverable_error, term()}}
+    | {error, term()}.
+do_query(
+    BridgeId,
+    Query,
+    ApplyMode,
+    #{poolname := PoolName, sql_templates := Templates} = State
+) ->
+    ?TRACE(
+        "SINGLE_QUERY_SYNC",
+        "sqlserver_connector_received",
+        #{query => Query, connector => BridgeId, state => State}
+    ),
+
+    %% only insert sql statement for single query and batch query
+    case apply_template(Query, Templates) of
+        {?ACTION_SEND_MESSAGE, SQL} ->
+            Result = ecpool:pick_and_do(
+                PoolName,
+                {?MODULE, worker_do_insert, [SQL, State]},
+                ApplyMode
+            );
+        {error, _} ->
+            Result = {error, {unrecoverable_error, failed_to_apply_sql_template}};
```

Comment:
Q: should we log the reason for the failure here?

Reply:
the reason will be always `failed_to_apply_sql_template` from `apply_template/2`